controller/lib/src/external_api/workload/controller.rs

```diff
@@ -1,46 +1,318 @@
-use super::{model::WorkloadInfo, service};
-use actix_web::{delete, get, patch, put, web, HttpResponse, Responder, Scope};
-
-pub fn get_services() -> Scope {
-    web::scope("/workload")
-        .service(get_workloads)
-        .service(get_workload)
-        .service(put_workload)
-        .service(patch_workload)
-        .service(delete_workload)
-}
+use crate::external_api::interface::ActixAppState;
 
-#[get("/")]
-pub async fn get_workloads() -> impl Responder {
-    service::get_workloads(&["1".to_string(), "2".to_string(), "3".to_string()]);
-    HttpResponse::Ok().body("workload")
-}
+use super::{model::Pagination, model::WorkloadDTO, model::WorkloadError, service};
+use actix_web::http::StatusCode;
+use actix_web::{web, HttpResponse, Responder, Scope};
 
-#[get("/{workload_id}")]
-pub async fn get_workload(workload_id: web::Path<String>) -> impl Responder {
-    service::get_workloads(&[workload_id.to_string()]);
-    HttpResponse::Ok().body(format!("workload {}", workload_id))
-}
+struct WorkloadController {}
 
-#[put("/")]
-pub async fn put_workload(body: web::Json<WorkloadInfo>) -> impl Responder {
-    let workload_info = body.into_inner();
-    service::create_workload(&workload_info);
-    HttpResponse::Ok().body("put_workload")
-}
+impl WorkloadController {
+    pub fn get_services(&self) -> Scope {
+        web::scope("/workload")
+            .service(
+                web::resource("/{namespace}/{workload_id}")
+                    .route(web::delete().to(WorkloadController::delete_workload))
+                    .route(web::get().to(WorkloadController::get_workload))
+                    .route(web::patch().to(WorkloadController::patch_workload)),
+            )
+            .service(
+                web::resource("/{namespace}")
+                    .route(web::put().to(WorkloadController::put_workload))
+                    .route(web::get().to(WorkloadController::get_all_workloads)),
+            )
+    }
+
+    /// `get_workload` is an async function that handle **/workload/\<namespace>/<workload_id>** route (GET)
+    /// # Description:
+    /// * Get a workload by id and namespace
+    /// # Arguments:
+    ///
+    /// * `params`: web::Path<(String, String)> - The first Path parameter is the namespace and the second the workload id.
+    pub async fn get_workload(
+        params: web::Path<(String, String)>,
+        data: web::Data<ActixAppState>,
+    ) -> impl Responder {
+        let (namespace, workload_id) = params.into_inner();
+
+        let mut workload_service = service::WorkloadService::new(&data.etcd_address)
+            .await
+            .map_err(|err| {
+                let mut http_response = HttpResponse::InternalServerError();
+                match err {
+                    WorkloadError::Etcd(msg) => http_response.body(format!("Etcd error: {} ", msg)),
+                    _ => http_response.body("Internal Server Error"),
+                }
+            })
+            .unwrap();
```

Comment:
No `unwrap`, please. Return a `Result` and propagate the error.

Reply:
In controller i can't propagate the error because it  needs to return HTTP response.
I use the unwrap because all error are handled line 43. If I replace unwrap by "?", the trait bound `actix_web::HttpResponse: ` is not satisfied ...